### ISSUE TITLE
Centos 7 PHP 8.1 openswoole 4.11 Segmentation fault when using composer  with openswoole enabled.

### ISSUE THREAD

**Issue by estagent** [openswoole/swoole-src#195]
Segmentation fault when using composer  with openswoole enabled.

$ composer --version
Composer version 2.2.9 2022-03-15 22:13:3

$ composer self-update
Loading composer repositories with package information
Segmentation fault

$ composer update
Loading composer repositories with package information
Segmentation fault

[ 2302.606382] php[22817]: segfault at 100000011 ip 00007f0ef21bec39 sp 00007ffcb2df47e0 error 4 in openswoole.so[7f0ef2072000+168000]
[ 2538.172864] php[23170]: segfault at 100000011 ip 00007fed669ccc39 sp 00007ffcffbe87a0 error 4 in openswoole.so[7fed66880000+168000]
[ 2591.233557] php[23219]: segfault at 17 ip 00007f23e0a02c39 sp 00007ffc9b20c3f0 error 4 in openswoole.so[7f23e08b6000+168000]
CentOS Linux release 7.9.2009 (Core)
3.10.0-1160.59.1.el7.x86_64 #1 SMP Wed Feb 23 16:47:03 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

openswoole

Open Swoole => enabled
Author => Open Swoole Group <hello@openswoole.com>
Version => 4.11.0
Built => Mar 24 2022 16:14:20
coroutine => enabled with boost asm context
trace_log => enabled
epoll => enabled
eventfd => enabled
signalfd => enabled
cpu_affinity => enabled
spinlock => enabled
rwlock => enabled
sockets => enabled
openssl => OpenSSL 1.0.2k-fips  26 Jan 2017
http2 => enabled
json => enabled
curl-native => enabled
pcre => enabled
c-ares => 1.10.0
zlib => 1.2.7
mutex_timedlock => enabled
pthread_barrier => enabled
futex => enabled
mysqlnd => enabled
async_redis => enabled
postgresql => enabled

Directive => Local Value => Master Value
swoole.enable_coroutine => On => On
swoole.enable_library => On => On
swoole.enable_preemptive_scheduler => Off => Off
swoole.display_errors => On => On
swoole.use_shortname => On => On
swoole.unixsock_buffer_size => 8388608 => 8388608

**Comment by yespire** [comment]
I'm experiencing the same segmentation fault on php81

If you don't need the new stuff from OpenSwoole, you can try the original Swoole package  https://pecl.php.net/package/swoole

I tried the Swoole-4.8.8, it is working okay without Segmentation Fault with the php81.
I have another box with php74, openswoole-4.10 is working fine there.

**Comment by estagent** [comment]
v4.11.0 worked when i ve configured with only  ./configure --enable-openssl --enable-sockets

**Comment by doubaokun** [comment]
@estagent @yespire 

We can't reproduce the issues you have seen, can you post more details: `cat /etc/os-release`, `php -v`, `php --ri openswoole`?

```
[root@ec6c5b84ddbf src]# cat /etc/os-release
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"
[root@ec6c5b84ddbf src]# php -v
PHP 8.1.5 (cli) (built: Apr 12 2022 16:19:58) (NTS gcc x86_64)
Copyright (c) The PHP Group
Zend Engine v4.1.5, Copyright (c) Zend Technologies
    with Zend OPcache v8.1.5, Copyright (c), by Zend Technologies
[root@ec6c5b84ddbf src]# php --ri openswoole

openswoole

Open Swoole => enabled
Author => Open Swoole Group <hello@openswoole.com>
Version => 4.11.0
Built => Apr 14 2022 21:52:00
coroutine => enabled with boost asm context
epoll => enabled
eventfd => enabled
signalfd => enabled
cpu_affinity => enabled
spinlock => enabled
rwlock => enabled
pcre => enabled
zlib => 1.2.7
mutex_timedlock => enabled
pthread_barrier => enabled
futex => enabled
async_redis => enabled

Directive => Local Value => Master Value
swoole.enable_coroutine => On => On
swoole.enable_library => On => On
swoole.enable_preemptive_scheduler => Off => Off
swoole.display_errors => On => On
swoole.use_shortname => On => On
swoole.unixsock_buffer_size => 8388608 => 8388608
[root@ec6c5b84ddbf src]# composer self-update
You are already using the latest available Composer version 2.3.5 (stable channel).
```

**Comment by XternalSoft** [comment]
Same problem on Fedora 35:

**cat /etc/os-release**
NAME="Fedora Linux"
VERSION="35 (Workstation Edition)"
ID=fedora
VERSION_ID=35
VERSION_CODENAME=""
PLATFORM_ID="platform:f35"
PRETTY_NAME="Fedora Linux 35 (Workstation Edition)"
ANSI_COLOR="0;38;2;60;110;180"
LOGO=fedora-logo-icon
CPE_NAME="cpe:/o:fedoraproject:fedora:35"
HOME_URL="https://fedoraproject.org/"
DOCUMENTATION_URL="https://docs.fedoraproject.org/en-US/fedora/f35/system-administrators-guide/"
SUPPORT_URL="https://ask.fedoraproject.org/"
BUG_REPORT_URL="https://bugzilla.redhat.com/"
REDHAT_BUGZILLA_PRODUCT="Fedora"
REDHAT_BUGZILLA_PRODUCT_VERSION=35
REDHAT_SUPPORT_PRODUCT="Fedora"
REDHAT_SUPPORT_PRODUCT_VERSION=35
PRIVACY_POLICY_URL="https://fedoraproject.org/wiki/Legal:PrivacyPolicy"
VARIANT="Workstation Edition"
VARIANT_ID=workstation

**php --ri openswoole**

openswoole

Open Swoole => enabled
Author => Open Swoole Group <hello@openswoole.com>
Version => 4.11.0
Built => Apr 23 2022 19:46:37
coroutine => enabled with boost asm context
epoll => enabled
eventfd => enabled
signalfd => enabled
cpu_affinity => enabled
spinlock => enabled
rwlock => enabled
sockets => enabled
openssl => OpenSSL 1.1.1n  FIPS 15 Mar 2022
dtls => enabled
json => enabled
curl-native => enabled
pcre => enabled
zlib => 1.2.11
brotli => E16777225/D16777225
mutex_timedlock => enabled
pthread_barrier => enabled
futex => enabled
mysqlnd => enabled
async_redis => enabled

Directive => Local Value => Master Value
swoole.enable_coroutine => On => On
swoole.enable_library => On => On
swoole.enable_preemptive_scheduler => Off => Off
swoole.display_errors => On => On
swoole.use_shortname => On => On
swoole.unixsock_buffer_size => 8388608 => 8388608


**composer self-update**
You are already using the latest available Composer version 2.3.5 (stable channel).


**composer outdated**
Info from https://repo.packagist.org: #StandWithUkraine
[1]    686188 segmentation fault (core dumped)  composer outdated

**Comment by doubaokun** [comment]
@XternalSoft please try master branch

**Comment by XternalSoft** [comment]
Hello @doubaokun ,

Same crash with master

Some tests  (compilation on each test):

OK with --enable-openssl 
OK with --enable-openssl --enable-mysqlnd
OK with --enable-openssl --enable-mysqlnd --enable-sockets
**KO** with --enable-openssl --enable-mysqlnd --enable-sockets --enable-swoole-curl

**Comment by XternalSoft** [comment]
**KO** with only --enable-swoole-curl

**Comment by doubaokun** [comment]
@XternalSoft more protections are added, please try the latest master branch.

**Comment by XternalSoft** [comment]
Segfault :( 

**php --ri openswoole**

openswoole

Open Swoole => enabled
Author => Open Swoole Group <hello@openswoole.com>
Version => 4.11.1-dev
Built => Apr 25 2022 09:14:02
coroutine => enabled with boost asm context
epoll => enabled
eventfd => enabled
signalfd => enabled
cpu_affinity => enabled
spinlock => enabled
rwlock => enabled
sockets => enabled
openssl => OpenSSL 1.1.1n  FIPS 15 Mar 2022
dtls => enabled
http2 => enabled
json => enabled
curl-native => enabled
pcre => enabled
c-ares => 1.17.2
zlib => 1.2.11
brotli => E16777225/D16777225
mutex_timedlock => enabled
pthread_barrier => enabled
futex => enabled
mysqlnd => enabled
async_redis => enabled

Directive => Local Value => Master Value
swoole.enable_coroutine => On => On
swoole.enable_library => On => On
swoole.enable_preemptive_scheduler => Off => Off
swoole.display_errors => On => On
swoole.use_shortname => On => On
swoole.unixsock_buffer_size => 8388608 => 8388608

**composer outdated**
Info from https://repo.packagist.org: #StandWithUkraine
[1]    63845 segmentation fault (core dumped)  composer outdated

**Comment by XternalSoft** [comment]
New idea : 
with **composer -vvv outdated** 

A see that core dump when composer look google apiclient package (i tested with and without composer cache)


```
Downloading https://repo.packagist.org/p2/google/apiclient-services.json
Downloading https://repo.packagist.org/p2/google/apiclient-services~dev.json
[200] https://repo.packagist.org/p2/google/apiclient-services.json
Writing /home/xxx/.cache/composer/repo/https---repo.packagist.org/provider-google~apiclient-services.json into cache
[1]    114255 segmentation fault (core dumped)  composer -vvv outdated
```

**Comment by doubaokun** [comment]
We are looking at this issue.

**Comment by doubaokun** [comment]
@XternalSoft there were some issues with old version libcurl on centos 7. Our tests have passed at our environment, Could you confirm?

**Comment by XternalSoft** [comment]
Hello @doubaokun ,

I also have a Centos 7 server but I was talking about a crash on my local PC which is on Fedora 35. 

**Fedora 35:**
libcurl-7.79.1-1.fc35.x86_64

**Centos 7:**
libcurl-7.29.0-59.el7_9.1.x86_64

But I retest with the master and tell you.

**Comment by XternalSoft** [comment]
I'm sorry but it's still the same :( 

**php --ri openswoole**

```
openswoole

Open Swoole => enabled
Author => Open Swoole Group <hello@openswoole.com>
Version => 4.11.1-dev
Built => Apr 26 2022 18:47:51
coroutine => enabled with boost asm context
epoll => enabled
eventfd => enabled
signalfd => enabled
cpu_affinity => enabled
spinlock => enabled
rwlock => enabled
sockets => enabled
openssl => OpenSSL 1.1.1n  FIPS 15 Mar 2022
dtls => enabled
http2 => enabled
json => enabled
curl-native => enabled
pcre => enabled
c-ares => 1.17.2
zlib => 1.2.11
brotli => E16777225/D16777225
mutex_timedlock => enabled
pthread_barrier => enabled
futex => enabled
mysqlnd => enabled
async_redis => enabled

Directive => Local Value => Master Value
swoole.enable_coroutine => On => On
swoole.enable_library => On => On
swoole.enable_preemptive_scheduler => Off => Off
swoole.display_errors => On => On
swoole.use_shortname => On => On
swoole.unixsock_buffer_size => 8388608 => 8388608
```

**composer outdated**
Info from https://repo.packagist.org: #StandWithUkraine
[1]    60363 segmentation fault (core dumped)  composer outdated


I thought it might be because of a large number of packages so I tested with a fresh install of Laravel and I have the same problem.

```
composer create-project laravel/laravel example-app
cd example-app
composer outdated
[1]    62705 segmentation fault (core dumped)  composer outdated

```

I will make another test, I will create a new machine under Fedora 35 to see if it is not related to my system which would be "polluted"  (if that can help you)

**Comment by doubaokun** [comment]
@XternalSoft thanks, can you also post the output of your core dump with “gdb php core”?

**Comment by XternalSoft** [comment]
Output

```
Core was generated by `php /usr/bin/composer outdated'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  zend_get_gc_buffer_add_zval (zv=0xc0dedbad, gc_buffer=0x1a29d98 <executor_globals+1624>) at /usr/local/php-fpm/8.0/include/php/Zend/zend_gc.h:99
99		if (Z_REFCOUNTED_P(zv)) {
```

Otherwise, I had a VM under AlmaLinux 8 that I use as a test. 
Everything works fine.
So I'm going to check with a new Fedora 35, it's just going to take a bit longer :p

**Comment by doubaokun** [comment]
@XternalSoft can you send the following info when seeing crash and also the `full output` of gdb:

```
php -v
gcc -v
php -m
```

**Comment by XternalSoft** [comment]
Hello,

I don't know how gdb works but here is the full output I get.
(by the way I just finished preparing a brand new VM with a Fedora Server 35 and I also have the crash)

**$ gdb php /tmp/php.core.861**
```
GNU gdb (GDB) Fedora 11.2-2.fc35
Copyright (C) 2022 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-redhat-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from php...

warning: Can't open file /dev/zero (deleted) during file-backed mapping note processing
[New LWP 861]

This GDB supports auto-downloading debuginfo from the following URLs:
https://debuginfod.fedoraproject.org/ 
Enable debuginfod for this session? (y or [n]) y
Debuginfod has been enabled.
To make this setting permanent, add 'set debuginfod enabled on' to .gdbinit.
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
Core was generated by `php /usr/sbin/composer outdated'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  zend_get_gc_buffer_add_zval (zv=0xc0dedbad, gc_buffer=0x1a29d98 <executor_globals+1624>)
    at /usr/local/php-fpm/8.0/include/php/Zend/zend_gc.h:99
99		if (Z_REFCOUNTED_P(zv)) {
(gdb) 

```

**$ php -v**
```
PHP 8.0.18 (cli) (built: Apr 27 2022 16:15:02) ( NTS )
Copyright (c) The PHP Group
Zend Engine v4.0.18, Copyright (c) Zend Technologies
    with Zend OPcache v8.0.18, Copyright (c), by Zend Technologies


```
**$ gcc -v**
```
Utilisation des specs internes.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/libexec/gcc/x86_64-redhat-linux/11/lto-wrapper
OFFLOAD_TARGET_NAMES=nvptx-none
OFFLOAD_TARGET_DEFAULT=1
Cible : x86_64-redhat-linux
Configuré avec: ../configure --enable-bootstrap --enable-languages=c,c++,fortran,objc,obj-c++,ada,go,d,lto --prefix=/usr --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=http://bugzilla.redhat.com/bugzilla --enable-shared --enable-threads=posix --enable-checking=release --enable-multilib --with-system-zlib --enable-__cxa_atexit --disable-libunwind-exceptions --enable-gnu-unique-object --enable-linker-build-id --with-gcc-major-version-only --with-linker-hash-style=gnu --enable-plugin --enable-initfini-array --with-isl=/builddir/build/BUILD/gcc-11.2.1-20220127/obj-x86_64-redhat-linux/isl-install --enable-offload-targets=nvptx-none --without-cuda-driver --enable-gnu-indirect-function --enable-cet --with-tune=generic --with-arch_32=i686 --build=x86_64-redhat-linux --with-build-config=bootstrap-lto --enable-link-serialization=1
Modèle de thread: posix
Algorithmes de compression LTO supportés: zlib zstd
gcc version 11.2.1 20220127 (Red Hat 11.2.1-9) (GCC) 

```

**$ php -m**
```
[PHP Modules]
bcmath
brotli
bz2
Core
ctype
curl
date
dom
exif
fileinfo
filter
gd
hash
iconv
igbinary
intl
json
libxml
mbstring
mysqli
mysqlnd
openssl
openswoole
pcntl
pcre
PDO
pdo_mysql
Phar
posix
redis
Reflection
session
SimpleXML
snappy
soap
sockets
SPL
standard
tokenizer
xml
xmlreader
xmlwriter
Zend OPcache
zip
zlib

[Zend Modules]
Zend OPcache

```

**Comment by doubaokun** [comment]
@XternalSoft Thanks for the support, we have fixed a issue on PHP8, please test the branch `fix_php8_fedora` on your machine if you can.

**Comment by XternalSoft** [comment]
It's works !!

Thanks ! 

I will test with a PHP8.1 and also on a Centos 7 to validate the whole :-)

**Comment by XternalSoft** [comment]
This works perfectly with PHP8.1.5 👍 (on Fedora 35)

**Comment by doubaokun** [comment]
All the issues should have been fixed since v4.11.1, thanks all for the feedback.
